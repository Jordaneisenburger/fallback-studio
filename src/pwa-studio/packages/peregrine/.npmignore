.eslintrc.js
/lib/**/__{docs,helpers,mocks,tests}__/**
/scripts
