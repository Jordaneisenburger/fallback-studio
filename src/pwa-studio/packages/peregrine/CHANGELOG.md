# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.5.0"></a>
# 0.5.0 (2018-08-13)


### Features

* multicasting REST API client ([#164](https://github.com/magento/pwa-studio/issues/164)) ([2852e14](https://github.com/magento/pwa-studio/commit/2852e14)), closes [#140](https://github.com/magento/pwa-studio/issues/140)
* Simulators for async testing ([#24](https://github.com/magento/pwa-studio/issues/24)) ([a380da9](https://github.com/magento/pwa-studio/commit/a380da9))
