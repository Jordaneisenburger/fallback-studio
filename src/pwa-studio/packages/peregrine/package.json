{
  "name": "@magento/peregrine",
  "version": "2.1.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "dist/index.js",
  "scripts": {
    "build": "concurrently --raw yarn:build:cjs yarn:build:esm",
    "build:cjs": "BABEL_ENV=production babel src --out-dir dist --root-mode 'upward' --source-maps",
    "build:dev": "yarn run build",
    "build:esm": "BABEL_ENV=development babel src --out-dir esm --root-mode 'upward' --source-maps",
    "clean": "rimraf dist esm",
    "prepublishOnly": "yarn run clean && yarn run build",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist",
    "watch": "yarn run build --watch"
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@babel/cli": "~7.2.3",
    "@babel/core": "~7.3.4",
    "@babel/plugin-proposal-class-properties": "~7.3.4",
    "@babel/plugin-proposal-object-rest-spread": "~7.3.4",
    "@babel/plugin-syntax-dynamic-import": "~7.2.0",
    "@babel/plugin-syntax-jsx": "~7.2.0",
    "@babel/plugin-transform-react-jsx": "~7.3.0",
    "@babel/plugin-transform-runtime": "~7.4.4",
    "@babel/runtime": "~7.4.2",
    "@storybook/react": "~4.1.13",
    "concurrently": "~4.1.0",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6",
    "rimraf": "~2.6.3"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "8.10.0 || >=10.7.0",
    "yarn": ">=1.12.0"
  },
  "module": "esm/index.js",
  "jsnext:main": "src/index.js",
  "es2015": "src/index.js",
  "sideEffects": false
}
