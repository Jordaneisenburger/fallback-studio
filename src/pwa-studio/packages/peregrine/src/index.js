import * as RestApi from './RestApi';
import * as Util from './util';

// hooks
export { useApolloContext } from './hooks/useApolloContext';
export { useEventListener } from './hooks/useEventListener';
export { useDropdown } from './hooks/useDropdown';
export { useQuery } from './hooks/useQuery';
export { useQueryResult } from './hooks/useQueryResult';
export { useSearchParam } from './hooks/useSearchParam';
export {
    WindowSizeContextProvider,
    useWindowSize
} from './hooks/useWindowSize';

// components
export { default as ContainerChild } from './ContainerChild';
export { default as List, Items, Item } from './List';
export { default as Page } from './Page';
export { default as Price } from './Price';
export { default as Router } from './Router';

// misc
export { RestApi };
export { Util };
export { default as createTestInstance } from './util/createTestInstance';
