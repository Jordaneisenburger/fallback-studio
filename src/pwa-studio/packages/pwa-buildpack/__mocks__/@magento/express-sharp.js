const middleware = jest.fn();
const expressSharp = (module.exports = jest.fn().mockReturnValue(middleware));
expressSharp.__mockMiddleware = middleware;
