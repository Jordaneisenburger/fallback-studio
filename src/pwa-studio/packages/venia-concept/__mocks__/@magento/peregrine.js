import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useApolloContext = jest.fn(peregrine.useApolloContext);
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useQuery = jest.fn(peregrine.useQuery);
const useQueryResult = jest.fn(peregrine.useQueryResult);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useApolloContext,
    useEventListener,
    useDropdown,
    useQuery,
    useQueryResult,
    useSearchParam,
    useWindowSize,

    useToasts
};
