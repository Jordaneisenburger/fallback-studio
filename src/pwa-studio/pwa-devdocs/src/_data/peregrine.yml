title: Peregrine
entries:
    - label: Overview
      url: /peregrine/

    - label: Routing in PWA Studio
      url: /peregrine/routing/

    - label: Utility modules
      entries:
          - label: REST API client
            url: /peregrine/reference/rest-api-client/

    - label: Custom React hooks
      entries:
          - label: useDropdown
            url: /peregrine/reference/hooks/useDropdown/

          - label: useEventListener
            url: /peregrine/reference/hooks/useEventListener/

          - label: usePagination
            url: /peregrine/reference/hooks/usePagination/

          - label: useSearchParam
            url: /peregrine/reference/hooks/useSearchParam/

          - label: useToasts
            url: /peregrine/reference/toasts/useToasts

          - label: useToastContext
            url: /peregrine/reference/toasts/useToastContext/

          - label: useWindowSize
            url: /peregrine/reference/hooks/useWindowSize/

    - label: React components
      entries:
          - label: ContainerChild
            url: /peregrine/reference/container-child/

          - label: Item
            url: /peregrine/reference/item/

          - label: Items
            url: /peregrine/reference/items/

          - label: List
            url: /peregrine/reference/list/

          - label: Price
            url: /peregrine/reference/price/

          - label: Router
            url: /peregrine/reference/router/
