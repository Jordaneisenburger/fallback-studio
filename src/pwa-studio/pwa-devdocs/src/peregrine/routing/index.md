---
title: Routing with PWA Studio
---

For web applications, routing is the process of mapping a URL to specific page resources.

In a multi-page application, routing is performed on the server side.
Every URL request fetches new HTML from the server and the browser loads the entire page.
This approach is inefficient because the same assets get loaded every time site navigation occurs.

For a single-page application (SPA), such as a progressive web app, routing is performed on the client side.
Single-page applications do not reload the browser during internal navigation.
Instead, the application uses the URL to fetch smaller pieces of data from the server and updates specific items on the page.

## Routing components

The following is a list of files provided by Peregrine to support routing:

| File                         | Description                                          |
| ---------------------------- | ---------------------------------------------------- |
| [`Router.js`][]              | Defines the **MagentoRouter** component              |
| [`MagentoRouteHandler.js`][] | Renders the correct component based on a given route |
| [`resolveUnknownRoute.js`][] | Gets the appropriate page type for a route           |
| [`webpackInterop.js`][]      | Load the Webpack chunk for a component               |
| [`Page.js`][]                | Sets up the MagentoRouteHandler as a router consumer |

## How it works

The following sections explain the purpose of each component when resolving a route and rendering content on a page.

### MagentoRouter

Routing starts in the **MagentoRouter** component.
It uses the [`react-router`][] library to implement a declarative approach to routing.
It also uses the [React Context][] API to create a `Consumer` and `Provider` pair.
These components allow the router to share information to any of its children without explicitly passing props down the tree.

Since routes can affect different pieces of a page, the MagentoRouter component should be located near the top of the React DOM tree.

### Page

The **Page** component sets up the **MagentoRouteHandler** as a child of a **MagentoRouter** `Consumer` component.
This lets the handler use routing information from it's MagentoRouter parent.

In the Venia storefront, the Page component exists alongside the Navigation and Shopping Cart components.
This lets the application shell update content inside the Page component during navigation without reloading data in the other parts of the page.

**Tip:**
Use the **Page** and **MagentoRouter** components to integrate the PWA Studio routing feature into your own PWA project.
{: .bs-callout .bs-callout-tip}

### MagentoRouteHandler

The **MagentoRouteHandler** component uses the route information, passed in by the MagentoRouter `Provider`, to determine which RootComponent to render.
It does this using the **resolveUnknownRoute** and **fetchRootComponent** helper components.

While the helper components resolves the route and gets the correct root component, the handler renders a loading message component.
When a root component is successfully loaded it uses the current application state to render the correct content on the page.
If an error occurs, such as a 404 error, the handler renders an error message component.

See [Loading and error message components][] to learn how to specify the loading and error message component.

#### resolveUnknownRoute

The **resolveUnknownRoute** helper component resolves a given route and provides information about its assigned RootComponent.

On a fresh page load, a DOM element with id `url-resolver` contains enough information for looking up the correct RootComponent chunk using the `roots-manifest.json` file.

The `roots-manifest.json` file contains information about the RootComponent chunks generated during a build.
The file itself is also generated during a build.

On subsequent navigations within the PWA, this component sends a request to the Magento backend asking for information about a specific route.

#### fetchRootComponent

The **fetchRootComponent** helper component uses Webpack to asynchronously load a specific RootComponent chunk.

## Loading and error message components

To specify the components to render during loading and error handling, specify the component that handles these two states as a child of the **Page** component.

The following code snippets from the Venia storefront project show how [App][] component sets the custom [ErrorView][] component as a child of Page.
This configures the Page component to use ErrorView for displaying the loading and error messages.

**`packages/venia-concept/src/components/ErrorView/errorView.js`**

```jsx
import React, { Component } from 'react';

const messages = new Map()
    .set('loading', 'Loading...')
    .set('notFound', '404 Not Found')
    .set('internalError', '500 Internal Server Error');

class ErrorView extends Component {
    render() {
        const { loading, notFound } = this.props;
        const message = loading
            ? messages.get('loading')
            : notFound
                ? messages.get('notFound')
                : messages.get('internalError');

        return <h1>{message}</h1>;
    }
}

export default ErrorView;
```

**`packages/venia-concept/src/components/App/app.js`**

```jsx
...
import { Page } from '@magento/peregrine';
...
import ErrorView from 'src/components/ErrorView';
...
const renderRoutingError = props => <ErrorView {...props} />;

class App extends Component {
...
    render(){
    ...
       return (
           <Fragment>
               <Main isMasked={overlay}>
                   <Page>{renderRoutingError}</Page>
               </Main>
               <Mask isActive={overlay} dismiss={closeDrawer} />
               <Navigation isOpen={navIsOpen} />
               <MiniCart isOpen={cartIsOpen} />
           </Fragment>
       );
    }
}
...
```

[`Router.js`]: https://github.com/magento-research/pwa-studio/blob/master/packages/peregrine/src/Router/Router.js
[`MagentoRouteHandler.js`]: https://github.com/magento-research/pwa-studio/blob/master/packages/peregrine/src/Router/MagentoRouteHandler.js
[`resolveUnknownRoute.js`]: https://github.com/magento-research/pwa-studio/blob/master/packages/peregrine/src/Router/resolveUnknownRoute.js
[`webpackInterop.js`]: https://github.com/magento-research/pwa-studio/blob/master/packages/peregrine/src/Router/webpackInterop.js
[`Page.js`]: https://github.com/magento-research/pwa-studio/blob/master/packages/peregrine/src/Page/Page.js
[`react-router`]: https://github.com/ReactTraining/react-router
[React Context]: https://reactjs.org/docs/context.html
[ErrorView]:t://github.com/magento-research/pwa-studio/blob/master/packages/venia-concept/src/components/ErrorView/errorView.js 
[App]: https://github.com/magento-research/pwa-studio/blob/master/packages/venia-concept/src/components/App/app.js
[Loading and error message components]: #loading-and-error-message-components
