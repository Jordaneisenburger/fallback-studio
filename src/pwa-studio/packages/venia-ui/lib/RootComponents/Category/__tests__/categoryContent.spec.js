import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import CategoryContent from '../categoryContent';

jest.mock('@magento/peregrine/lib/context/app', () => {
    const state = {};
    const api = {
        toggleDrawer: jest.fn()
    };
    const useAppContext = jest.fn(() => [state, api]);
    return { useAppContext };
});

jest.mock('../../../components/Head', () => ({
    HeadProvider: ({ children }) => <div>{children}</div>,
    Title: () => 'Title'
}));

jest.mock('../../../components/Breadcrumbs', () => 'Breadcrumbs');
jest.mock('../../../components/Gallery', () => 'Gallery');
jest.mock('../../../components/Pagination', () => 'Pagination');

const classes = {
    root: 'a',
    title: 'b',
    gallery: 'c',
    pagination: 'd'
};

const data = {
    category: {
        name: 'Name',
        description: 'test'
    },
    products: {
        items: {
            id: 1
        }
    }
};

test('renders the correct tree', () => {
    const instance = createTestInstance(
        <CategoryContent
            pageControl={{}}
            data={data}
            pageSize={6}
            classes={classes}
        />
    );

    expect(instance.toJSON()).toMatchSnapshot();
});
