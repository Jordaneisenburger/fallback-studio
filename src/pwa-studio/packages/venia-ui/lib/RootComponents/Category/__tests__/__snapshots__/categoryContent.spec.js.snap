// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
Array [
  <Breadcrumbs />,
  "Title",
  <article
    className="a"
  >
    <h1
      className="b"
    >
      <div>
        Name
      </div>
    </h1>
    <section
      className="c"
    >
      <Gallery
        items={
          Object {
            "id": 1,
          }
        }
      />
    </section>
    <div
      className="d"
    >
      <Pagination
        pageControl={Object {}}
      />
    </div>
  </article>,
]
`;
